includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -485,11 +490,50 @@ public function process_payment_for_order( $args ) {
 
 			$intent_id = $intent->get_id();
 			$status    = $intent->get_status();
+			$charge_id = $intent->get_charge_id();
+		} else {
+			// For $0 orders, we need to save the payment method using a setup intent.
+			$intent = $this->payments_api_client->create_setup_intent(
+				$payment_method,
+				$customer_id,
+				'true'
+			);
+
+			$intent_id = $intent['id'];
+			$status    = $intent['status'];
+			$charge_id = '';
 
+			// In SCA cases the setup intent status might be requires_action and we should display the authentication modal.
+			// For now, since we're not supporting SCA cards, we can ignore that status.
+			if ( 'succeeded' !== $status ) {
+				throw new Exception( __( 'Failed to add the provided payment method. Please try again later', 'woocommerce-payments' ) );
+			}
+		}
+
+		if ( ! empty( $intent ) ) {
 			if ( 'succeeded' !== $status && 'requires_capture' !== $status ) {
 				$intent_failed = true;
 			}
 
+			if ( $save_payment_method && ! $intent_failed ) {
+				try {
+					$token = $this->token_service->add_payment_method_to_user( $payment_method, $user );
+				} catch ( Exception $e ) {
+					// If saving the token fails, log the error message but catch the error to avoid crashing the checkout flow.
+					Logger::log( 'Error when saving payment method: ' . $e->getMessage() );
+				}
+			}
+
+			$order_tokens = $order->get_payment_tokens();
+			if ( ! empty( $token ) && $token->get_id() !== end( $order_tokens ) ) {
+				$order->add_payment_token( $token );
+				// Set payment token for subscriptions, so it can be used for renewals.
+				$subscriptions = wcs_get_subscriptions_for_order( $order_id );
+				foreach ( $subscriptions as $subscription ) {
+					$subscription->add_payment_token( $token );
+				}
```

Comment:
I left a question on whether this is really necessary in https://github.com/Automattic/woocommerce-payments/pull/826#discussion_r476730272.

Reply:
Resolving this as it was necessary and the commits were `cherry-picked` in #826 (https://github.com/Automattic/woocommerce-payments/pull/826/commits/0fe09c2c9e0c4e1eb976f8b2acabdd06c3be5f43 and https://github.com/Automattic/woocommerce-payments/pull/826/commits/1b3692bd7975dea897c462178c2ebc7694a0e94e).